### ISSUE TITLE
selection.append seems case-sensitive (non-svg)

### ISSUE THREAD

**Issue by martingraham** [mbostock/d3#2722]
See https://jsfiddle.net/16n6uxqg/3/

Appending some "th" / "TH" elements, I noticed they seem to get added ok whatever the case, but then don't get picked up by either .selectAll("TH") or .selectAll("th") if originally appended using uppercase - "TH". The styling's different too even though the inspector says they're th elements. They do get picked up using .selectAll("*")

Elements I add directly as "TH" using document appendChild are picked up ok by d3's .selectAll. So are hard-coded upper case "TH" elements in HTML.

I'm seeing this in chrome 48 / firefox 43 / opera 34

These aren't inside SVG elements so it's not the webkit camelcase issue that's been reported in the past.

Am I missing something obvious (I wouldn't be surprised)?

**Comment by mbostock** [comment]
Nice catch. I’m kind of amazed this hasn’t been reported before, but my guess is that most people have been passing lower-case names to *selection*.append.

The problem here is an interaction between [document.createElement](https://developer.mozilla.org/en-US/docs/Web/API/Document/createElement),  [document.createElementNS](https://developer.mozilla.org/en-US/docs/Web/API/Document/createElementNS) and *selection*.append. 

According to the spec, createElement has special behavior on HTML documents:

> When called on a document object flagged as an HTML document, createElement() lower-cases its argument prior to creating the element.

This behavior does not apply to createElementNS. And, since *selection*.append inherits the namespaceURI from the parent element, it almost always uses createElementNS, even for HTML documents. So, when you append th elements, instead of creating them like so:

```
document.createElement("th");
```

It is using this:

```
document.createElementNS("http://www.w3.org/1999/xhtml", "th");
```

So, if you say "TH" rather than "th", it creates the wrong element type (and hence why the table doesn’t render correctly). Confusingly, though, the DOM spec actually *upper-cases* tag names when you inspect them via *element*.tagName, so it’s not immediately obvious that something went wrong. However, if you pull out the body innerHTML, you can see the top row is unusual:

```html
<div class="holder">
  <table>
    <tbody><tr id="d3RowUpper"><TH>HEY</TH><TH>YOU</TH></tr>
    <tr id="d3RowLower"><th>WHO</th><th>ME</th></tr>
    <tr id="docRow"><th>YEAH</th><th>YOU</th></tr>
    <tr id="htmlRow">
      <th>WELL</th>
      <th>WHAT?</th>
    </tr>
  </tbody></table>
</div>
```

The fix here is probably to have D3 lower-case the specified element name if the namespace is XHTML. Alternatively, D3 could use document.createElement rather than document.createElementNS if the parent element’s namespace matches the document element’s namespace. The latter should be equivalent, though perhaps slightly more general.

**Comment by mbostock** [comment]
Fixed in 3.5.14.
